crates/provider/src/utils.rs

```diff
@@ -138,6 +142,31 @@ pub(crate) fn convert_to_hashes<BlockResp: alloy_network::BlockResponse>(
     })
 }
 
+/// Fetches full blocks for a list of block hashes
+pub(crate) async fn hashes_to_blocks<BlockResp: BlockResponse + RpcRecv>(
+    hashes: Vec<B256>,
+    client: WeakClient,
+    full: bool,
+) -> TransportResult<Vec<Option<BlockResp>>> {
+    let mut futures = Vec::with_capacity(hashes.len());
+
+    for hash in hashes {
+        let fut = {
+            let client =
+                client.upgrade().ok_or(TransportError::local_usage_str("client dropped"))?;
+
+            client
+                .request::<_, Option<BlockResp>>("eth_getBlockByHash", (hash, full))
+                .map_resp(|resp| if !full { convert_to_hashes(resp) } else { resp })
+        };
+
+        futures.push(fut);
+    }
+
+    let blocks = futures::future::try_join_all(futures).await?;
```

Comment:
this takes an iterator so all of this function can be just `try_join_all(hashes.into_iter().map(|hash| ...))`